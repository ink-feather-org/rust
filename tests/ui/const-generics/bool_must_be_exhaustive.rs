#![feature(generic_const_exprs)]
#![feature(impl_unified_exhaustive_const_traits)]
#![allow(incomplete_features)]

pub const fn bool_to_usize(b: bool) -> usize {
    b as usize
}

pub struct ConstOption<T, const S: bool> where [T; bool_to_usize(S)]:, {
    _v: [T; bool_to_usize(S)]
}

impl<T: Default> Default for ConstOption<T, true> {
    fn default() -> Self {
        Self {
        _v: [T::default()]
        }
    }
}

fn _test_func<const N: usize>() where ConstOption<usize, {N >= 5}>: Default,
    [usize; bool_to_usize(N >= 5)]: {

}

#[derive(Default)]
pub struct Arg<const N: usize> where [(); bool_to_usize(N <= 0)]:, [(); bool_to_usize(N <= 1)]:, {
    _a: ConstOption<usize, { N <= 0 }>,
    //~^ ERROR trait bound
    _b: ConstOption<usize, { N <= 1 }>,
    //~^ ERROR trait bound
}

fn main() {
  let _def = Arg::<2>::default();
}
