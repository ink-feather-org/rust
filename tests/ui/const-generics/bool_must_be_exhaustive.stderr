error[E0277]: the trait bound `ConstOption<usize, { N <= 0 }>: Default` is not satisfied
  --> $DIR/bool_must_be_exhaustive.rs:28:5
   |
LL | #[derive(Default)]
   |          ------- in this derive macro expansion
LL | pub struct Arg<const N: usize> where [(); bool_to_usize(N <= 0)]:, [(); bool_to_usize(N <= 1)]:, {
LL |     _a: ConstOption<usize, { N <= 0 }>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `ConstOption<usize, { N <= 0 }>`
   |
   = help: the trait `Default` is implemented for `ConstOption<T, true>`
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `ConstOption<usize, { N <= 1 }>: Default` is not satisfied
  --> $DIR/bool_must_be_exhaustive.rs:30:5
   |
LL | #[derive(Default)]
   |          ------- in this derive macro expansion
...
LL |     _b: ConstOption<usize, { N <= 1 }>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `ConstOption<usize, { N <= 1 }>`
   |
   = help: the trait `Default` is implemented for `ConstOption<T, true>`
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
